user_sync/connector/directory_ldap.py

```diff
@@ -269,35 +272,55 @@ def iter_users(self, users_filter):
                 if (last_attribute_name != None):
                     self.logger.warn('No email attribute: %s for dn: %s', last_attribute_name, dn)
                 continue
+
+            source_attributes = {}
             
             user = user_sync.connector.helper.create_blank_user()
             user['email'] = email
+            source_attributes['email'] = email
                 
             username, last_attribute_name = self.user_username_formatter.generate_value(record)
             if (username == None and last_attribute_name != None):
                 self.logger.info('No username attribute: %s for dn: %s', last_attribute_name, dn)    
             user['username'] = username if username != None else email
+            if (username != None):
```

Comment:
You're actually making two different points here. (1) You want username to be overridden with email if not present. I definitely disagree with this. It loses information, and what you're looking for already exists in target_attributes. (2) You want all attributes we know about to set to None if not present. I reacted negatively to this at first to, but as I try to figure out why, I think it's because it makes the log harder to scan. But maybe that's not a good enough reason. (Yeah, yeah, I know, you're going to say that's about us, not about them. But actually I think they'll be living in the logs too...)